### ISSUE TITLE
growing arrays (slicearray, but probably many others)

### ISSUE THREAD

**Issue by gesellkammer** [csound/csound#1071]
This seems a related issue to #1070. When calling slicearray in the same instrument with different arguments, the resulting array is always as big as the biggest call until now. The array seems to be taken from a previous instance and it's size is never reset (maybe a problem with `arrayensure`?). Here is some code to test this. The second time instr 1 is scheduled, the size of the array returned by slicearray is wrong. If the order of the score lines is inverted, everything is as it should. 

```csound

<CsoundSynthesizer>
<CsOptions>
; -odac           
; -iadc     
; -d     ;;;RT audio I/O
;; -+rtaudio=jack
</CsOptions>

<CsInstruments>
sr     = 44100
ksmps  = 64
nchnls = 2
0dbfs  = 1

instr 1
  iend = p4
  iFoo[] fillarray 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10
  iBar[] slicearray iFoo, 0, iend
  ilen lenarray iBar  
  printarray iBar
  print ilen
  turnoff
endin

</CsInstruments>

<CsScore>

i1 0   1 10
i1 0.1 1 4

</CsScore>
</CsoundSynthesizer>
```

**Comment by jpffitch** [comment]
Unrlated to #1070

**Comment by jpffitch** [comment]
Looks like a local variable is inheriting an old allocation as if it were global.

**Comment by jpffitch** [comment]
Indeed the value of iBar is the old version with its data.  So this is Steven's code on allocation of variables.

**Comment by vlazzarini** [comment]
I think iBarr would inherit normally as per usual rules, we don't realloc if memory exists for an instance.

I don't know what's the right thing in this case.

**Comment by vlazzarini** [comment]
actually, no. There are two separate allocations of instr 1 as they overlap. There should be two distinct instances.

**Comment by jpffitch** [comment]
I culd not find where i allocates or does not allocate new space.  My 
currnt tesrt is
i1 0.1 0.1 4
i1 0.2   0.1 10
i1 0.3 0.1 4
which shows clearly 9via gdb) tha iBar has an array allocated of 
none zero length in the second call to slicearray, and so if the array is 
long enough it will reuse it and maintain the size.  Thia is wrong.  It 
should reset the local array before init cycle starts.

**Comment by vlazzarini** [comment]
In insert.c, new memory space for instr variables is allocated, but the actual array space is
allocated by the init time function of the opcode (tabensure etc).

**Comment by jpffitch** [comment]
The array is not being allocated properly in insert.c then as before the 
variable is ever used I would expect it to be an array with NULL data and 
that is not the case

new alloc for instr 1:

Breakpoint 1, tabslice (csound=0x61f360, p=0x869408)
     at /home/jpff/csound6/Opcodes/arrays.c:2416
2416    static int32_t tabslice(CSOUND *csound, TABSLICE *p) {
(gdb) print *p
$1 = {h = {nxti = 0x869468, nxtp = 0x0, iopadr = 0x7ffff79bfd20 
<tabslice>,
     opadr = 0x0, optext = 0x8562b0, insdshead = 0x865150}, tab = 0x865348,
   tabin = 0x865310, start = 0x854118, end = 0x865300, inc = 0x854638, len 
= 0}
(gdb) print *p->tab
$2 = {dimensions = 0, sizes = 0x0, arrayMemberSize = 0,
   arrayType = 0x7ffff7d9e620 <CS_VAR_TYPE_I>, data = 0x0, allocated = 0}
(gdb) c
Continuing.
   0.0000 1.0000 2.0000 3.0000 4.0000
instr 1:  ilen = 5.000
B  0.100 ..  0.200 T  0.200 TT  0.200 M:  0.00000  0.00000

Breakpoint 1, tabslice (csound=0x61f360, p=0x869408)
     at /home/jpff/csound6/Opcodes/arrays.c:2416
2416    static int32_t tabslice(CSOUND *csound, TABSLICE *p) {
(gdb) print *p
$3 = {h = {nxti = 0x869468, nxtp = 0x0, iopadr = 0x7ffff79bfd20 
<tabslice>,
     opadr = 0x0, optext = 0x8562b0, insdshead = 0x865150}, tab = 0x865348,
   tabin = 0x865310, start = 0x854118, end = 0x865300, inc = 0x854638, len 
= 0}
(gdb) print *p->tab
$4 = {dimensions = 1, sizes = 0x837570, arrayMemberSize = 8,
   arrayType = 0x7ffff7d9e620 <CS_VAR_TYPE_I>, data = 0x78b870, allocated = 
40}
(gdb)

This is from the test program.  In the second instance of i1 the first 
call to tabslice has a valid array of a size already allocated.  In this 
case the second call wants a larger array so it is realloced by tabensure. 
In the third i1 the preallocated array is big enough for the tabensure to 
say all is OK.  Previously you needed this to be the case that size does 
not reduce, but now we have a case where that is "wrong".

I think the solution  should be in the allocation of the instance, 
clearing arrays.

**Comment by vlazzarini** [comment]
I would not expect it to be a NULL array because:

1. The instance is reused and the memory is not reallocated or reset (that is the behaviour I would expect).

2. The opcode init function does not reallocate the array because it has enough memory. That is also what I'd expect. 

I am not sure I understand this 

"Previously you needed this to be the case that size does
not reduce, but now we have a case where that is "wrong"."

Why is it wrong?

**Comment by jpffitch** [comment]
"Wrong" because the behaviour of an array in an instrument is undetermined 
by that insntance but the final state of some previous instance of that 
instrument, and howa local variable was left.  'this breaks the localness 
of variables for a start and can give rise to differen resukts.

Would you expect these two fragments to produce different results?

1:

   iend = p4
   iFoo[] fillarray 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10
   iBar[] slicearray iFoo, 0, iend
   ilen lenarray iBar
   print ilen


2:

   iend = p4
   iFoo[] fillarray 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10
   iBar[] init 1
   iBar slicearray iFoo, 0, iend
   ilen lenarray iBar
   printarray iBar
   print ilen

They do

**Comment by vlazzarini** [comment]
But you are not suggesting we reallocate the memory that already exists and has enough size each time, are you?
What I would expect is that tabensure could maybe just mark the array with the requested length, and leave the capacity unchanged so it can grow later with no reallocation.

**Comment by jpffitch** [comment]
There is code  to keep the capacity unchanged but fix the perceived size. 
You had an example in the pvs area for which that did not work so we left 
it at the current compromise.

If the allocation of localmemory trimmed the array to 0 then I would be 
happy as it would not have these difficult to explain oddities.  At 
present it seems that arrays are tied ven the rest of the instance is not.

I looked for where this could be done but I find insert.c rather opaque.

**Comment by jpffitch** [comment]
There is no leak in what I suggest and no inefficiency

**Comment by gesellkammer** [comment]
Is there something preventing solving this issue? This makes ALL array opcodes return incorrect results  when an instrument is reused

**Comment by kunstmusik** [comment]
(Issue seems serious, let's resolve this for 6.13 before it goes out.)

**Comment by jpffitch** [comment]
Steven, what do you think should be done?  Previous attempts failed to run a pvs example -- need to see if I can find which.  In some way slicearray should ensure the size of the answer not the size of the input array but that would seem to require doing it twice or some nasty calculation....

**Comment by gesellkammer** [comment]
After analyzing the code I see that the bug here is just another incarnation of the problems seen when using an array more than once inside an instrument.

What happens here, as I understand it, is:

* when an instrument is reused, the variables are already initialized from the previous instance
* the second time the array variable is used, tabensure checks its size. If it needs to make it bigger, everything is fine
* if the opposite happens, the array is left with an incorrect perceived size

What needs to be done:

* at init-time, first we check that the array has been initialized.
* at init-time, the array is set to the desired size, either growing the array or setting the perceived size. This is done for every i-time operation, since the i-time size should be correct at any time
* at k-time (this is only valid for k-rate opcodes), the array's size is ensured to be the desired size (again either by growing the array if it is too small, or by adjusting the perceived size if it is too big)

Reallocation is only ever done to grow the array, otherwise only the perceived size is adjusted.

**Comment by jpffitch** [comment]
Can you articulate how what you suggest differ from the current code?  It 
looks like what I thought happens.

Sorry for being dense

**Comment by gesellkammer** [comment]
currently tabensure only adjusts perceived size if array is not 
initialized. Here is the code

static inline void tabensure(CSOUND *csound, ARRAYDAT *p, int size)
{
     if (p->data==NULL || p->dimensions == 0 ||
         (p->dimensions==1 && p->sizes[0] < size)) {
       size_t ss;
       if (p->data == NULL) {
         CS_VARIABLE* var = p->arrayType->createVariable(csound, NULL);
         p->arrayMemberSize = var->memBlockSize;
       }
       ss = p->arrayMemberSize*size;
       if (p->data==NULL) {
         p->data = (MYFLT*)csound->Calloc(csound, ss);
         p->allocated = ss;
       }
       else if (ss > p->allocated) {
         p->data = (MYFLT*) csound->ReAlloc(csound, p->data, ss);
         p->allocated = ss;
       }
       if (p->dimensions==0) {
         p->dimensions = 1;
         p->sizes = (int32_t*)csound->Malloc(csound, sizeof(int32_t));
       }
       p->sizes[0] = size;
     }
     //p->sizes[0] = size;
}

p->sizes[0] = size only takes effect if array is not initialized. My suggestion is this:

This can only be called at init time, makes sure that array is initialized
and that it can allocate at least size elements. Then it ensures that the perceived
size is size

statuc inline void
tabensure_init(CSOUND *csound, ARRAYDAT *p, int size)
{
     size_t ss;
     if (p->dimensions==0) {
         p->dimensions = 1;
         p->sizes = (int32_t*)csound->Malloc(csound, sizeof(int32_t));
     }
     if (p->data == NULL) {
         CS_VARIABLE* var = p->arrayType->createVariable(csound, NULL);
         p->arrayMemberSize = var->memBlockSize;
         ss = p->arrayMemberSize*size;
         p->data = (MYFLT*)csound->Calloc(csound, ss);
         p->allocated = ss;
     } else if( (ss = p->arrayMemberSize*size) > p->allocated) {
         p->data = (MYFLT*) csound->ReAlloc(csound, p->data, ss);
         p->allocated = ss;
     }
     p->sizes[0] = size;
}

This can be called at perf-time, assumes that tabensure_init was called at init time.

static inline void
tabensure_perf(CSOUND *csound, ARRAYDAT *p, int size)
{
     if(UNLIKELY(p->sizes[0] != size)) {
         size_t ss = p->arrayMemberSize*size;
         if (ss > p->allocated) {
             p->data = (MYFLT*) csound->ReAlloc(csound, p->data, ss);
             p->allocated = ss;
         }
         p->sizes[0] = size;
     }
}

**Comment by kunstmusik** [comment]
I'm wondering if we shouldn't have gone all the way earlier with
implementing the arrays as arraylists.  I think this was going to be punted
to CS7, but seems like the issues now require moving that up.  We should
probably think about growing arrays via doubling and tracking both internal
array size and data array size (size according to user code).  We could
then do #513.

I'll take a break from merging develop and parser3 and look at this.

**Comment by gesellkammer** [comment]
allocated size and size according to user code are already tracked, the 
first in the `allocated` member and the second in `sizes[i]` (0 if array 
is 1D)

Of course reallocation could be done in a more efficient way, but at the 
very least the behaviour should be correct.

**Comment by jpffitch** [comment]
But we do track both internal array size and data array size.

**Comment by kunstmusik** [comment]
Sorry, I see it now that I'm on my computer. I agree with @gesellkammer 's analysis.  I'd be up for CS6:

* update arrays.h to have the two tabensures
* update code currently using old tabensure

For CS7, I'm thinking:

1. Update for size doubling (rather than realloc to given size)
2. Implement list operations (append, splice, shift, etc.)

We also have a lot of tabensure implemented but #ifdef'd out. Would be nice if those are removed now at this point too. 

We still need to consider changes when dimensions > 1 (probably will have to plan to wipe data in those situations to be safe).

**Comment by gesellkammer** [comment]
Since implementing this change touches all the code which uses arrays, maybe it would help to clarify the behaviour. In general, in any array operation which outputs an array doing some element-wise operation on an input array, the output array should have the same shape / size as the input array.

For example, in many arithmetic operations with arrays, the size of the resulting array depends on the left hand array, which is very surprising.

```
static int32_t tabarithset(CSOUND *csound, TABARITH *p)
{
    if (LIKELY(p->left->data && p->right->data)) {
      int32_t size;
      if (UNLIKELY(p->left->dimensions!=1 || p->right->dimensions!=1))
        return
          csound->InitError(csound, "%s",
                            Str("Dimensions do not match in array arithmetic"));
      /* size is the smallest of the two */
      size = p->left->sizes[0] < p->right->sizes[0] ?
        p->left->sizes[0] : p->right->sizes[0];
      tabensure(csound, p->ans, size);
      p->ans->sizes[0] = size;
      return OK;
    }
    else return csound->InitError(csound, "%s",
                                  Str("array-variable not initialised"));
}
```
There are many problematic / unclear cases in arrays.c, where the size used to call tabensure is wrong

* tabgen: two calls to tabensure?
* ftab2tabi / ftab2tab: the correct size is only set if the array was not previous initialized, otherwise the previous size is kept
* tabmap_set: the same problem, the size depends on the previous size of the array

See also Opcodes/arrays.c/tabgen (two calls to tabensure?)

**Comment by gesellkammer** [comment]
Since implementing this change touches all the code which uses arrays, 
maybe it would help to clarify the behaviour. In general, in any array 
operation which outputs an array doing some element-wise operation on an 
input array, the output array should have the same shape / size as the 
input array.

For example, in many arithmetic operations with arrays, the size of the 
resulting array depends on the left hand array, which is very surprising.

|static int32_t tabarithset(CSOUND *csound, TABARITH *p) { if 
(LIKELY(p->left->data && p->right->data)) { int32_t size; if 
(UNLIKELY(p->left->dimensions!=1 || p->right->dimensions!=1)) return 
csound->InitError(csound, "%s", Str("Dimensions do not match in array 
arithmetic")); /* size is the smallest of the two */ size = 
p->left->sizes[0] < p->right->sizes[0] ? p->left->sizes[0] : 
p->right->sizes[0]; tabensure(csound, p->ans, size); p->ans->sizes[0] = 
size; return OK; } else return csound->InitError(csound, "%s", 
Str("array-variable not initialised")); } |

There are many problematic / unclear cases in arrays.c, where the size 
used to call tabensure is wrong

  * tabgen: two calls to tabensure?
  * ftab2tabi / ftab2tab: the correct size is only set if the array was
    not previous initialized, otherwise the previous size is kept
  * tabmap_set: the same problem, the size depends on the previous size
    of the array

**Comment by jpffitch** [comment]
The size must depend on the size of the inputs left and right.  I d not 
follow your argument.

**Comment by gesellkammer** [comment]
The remark about `tabarithset` was wrong, I'm sorry, I misread the code. 

On the other hand there are major fixes needed if we accept that arrays can change size during performance. Currently all arithmetic methods do the following:

* they ensure the size of output array at init time
* at perf time they rely on that size still being correct
* they derive the size of the operation to perform on the size of the **answer** array, which might be wrong

See for example `tabmult` in arrays.c which is responsible for multiplying two k-arrays. If an array can change its size at perf-time, then the size of the operation must be calculated and ensured at perf time.

In this case `tabmult` should read 

```
static int32_t tabmult(CSOUND *csound, TABARITH *p)
{
    ARRAYDAT *ans = p->ans;
    ARRAYDAT *l   = p->left;
    ARRAYDAT *r   = p->right;
    int32_t lsize = p->left->sizes[0], rsize = p->right->sizes[0];
    int32_t size = lsize < rsize ? lsize : rsize;
    tabensure_perf(csound, p->ans, size);
    int32_t i;
    /* this can't happen so no need to check 
    if (UNLIKELY(ans->data == NULL || l->data== NULL || r->data==NULL))
      return csound->PerfError(csound, &(p->h),
                               Str("array-variable not initialised"));
    */
    for (i=0; i<size; i++)
      ans->data[i] = l->data[i] * r->data[i];
    return OK;
}
```

**Comment by jpffitch** [comment]
> See for example tabmult in arrays.c which is responsible for multiplying two
> k-arrays. If an array can change its size at perf-time, then the size of the
> operation must be calculated and ensured at perf time.


Yes; my error

I am unconvinced that we cannot get an uninitialised array.  Injudicious 
use of if - else - endif has been known to cause problems esewhere.  Note 
the tests usually labelled "RWD fix"

>     /* this can't happen so no need to check
>     if (UNLIKELY(ans->data == NULL || l->data== NULL || r->data==NULL))
>       return csound->PerfError(csound, &(p->h),
>                                Str("array-variable not initialised"));
>     */

**Comment by gesellkammer** [comment]
But those should never be NULL (ans is checked at init time in 
tabensure_init, the rest are right hand side vars, so if they are null 
this is a bug, not something that should be checked at runtime. But it 
is not a bad idea to leave the checks while the changes are being 
introduced, they can always be removed later.

**Comment by kunstmusik** [comment]
I'm looking at the code and wondering if there isn't a simpler way to solve this by using arrayInitMemory to always reset all sizes to 0.  Another way to simplify this is limit the number of dimensions to a fixed limit (say, 6 dimensions? I don't imagine anyone uses more than 3 today, do they?) For CS6, we can keep backwards compatibility by maintaining the existing ARRAYDAT structure and just preallocate sizes to hold 6 dimensions max. I think the clearing during initialization is what is missing. 

I'll try some things out now in a new branch and push for review if it works out.

**Comment by kunstmusik** [comment]
@gesellkammer @jpffitch @vlazzarini I made a change in the array-init-change branch (see commit 29014e1). 

What it does is modify tabensure to be like @gesellkammer's tabensure_init. I'm thinking that is the safest to have it run through all of this at init and perf-time both, since we don't know the size of input arrays to opcodes until perf-time and we may have opcodes that only run at perf-time. 

The original test code reports correct values here after this change. Could you all take a look and see if I've missed anything?  (I did do some additional changes for arrayInitMemory to just prealloc 6 dimensions of size as well.)

**Comment by gesellkammer** [comment]
Thanks for this @kunstmusik. 

I like the idea of having preallocated sizes. Regarding using the same version of tabensure both at init and performance time is ok also. Why are you double checking `p->data==NULL`? Can't you just unify the ifs?
Also to clarify: tabensure is intended to work ONLY on 1D arrays (it ensures that size[0] == size). In that case it should also ensure that `dimensions == 1` (so the check `dimensions == 0` should be changed to `dimensions != 1`. Below is the original and after that my version of it.

```c

static inline void tabensure(CSOUND *csound, ARRAYDAT *p, int size)
{
  size_t ss;
  if (p->data == NULL) {
    CS_VARIABLE* var = p->arrayType->createVariable(csound, NULL);
    p->arrayMemberSize = var->memBlockSize;
  }
  ss = p->arrayMemberSize*size;
  if (p->data==NULL) {
    p->data = (MYFLT*)csound->Calloc(csound, ss);
    p->allocated = ss;
  }
  else if (ss > p->allocated) {
    p->data = (MYFLT*) csound->ReAlloc(csound, p->data, ss);
    p->allocated = ss;
  }
  if (p->dimensions==0) {   
    p->dimensions = 1;
  }
  p->sizes[0] = size;
}

static inline void
tabensure(CSOUND *csound, ARRAYDAT *p, int size) {
    size_t ss = p->arrayMemberSize*size;
    if (p->data == NULL) {
        CS_VARIABLE* var = p->arrayType->createVariable(csound, NULL);
        p->arrayMemberSize = var->memBlockSize;
        p->data = (MYFLT*)csound->Calloc(csound, ss);
        p->allocated = ss;
    } else if(ss > p->allocated) {
        p->data = (MYFLT*) csound->ReAlloc(csound, p->data, ss);
        p->allocated = ss; }
    p->sizes[0] = size;
    p->dimensions = 1;
} 

```

**Comment by kunstmusik** [comment]
@gesellkammer That makes sense and I updated the code to the implementation you gave:

https://github.com/csound/csound/commit/c9b7f6b418bcea43c840928a1b227126c5a5dae0

I also added some documentation so that we have the behavior specified for future reference.

**Comment by vlazzarini** [comment]
We cannot have re-allocation at perf-time, is that going to happen?

**Comment by jpffitch** [comment]
Can someone try the new-arrays branch?  Does allocation at init time and does not change sizes

**Comment by rorywalsh** [comment]
Sorry John, won't be in a position to test until the middle of next week..

**Comment by gesellkammer** [comment]
I just took a look at arrays.h and this implementation does not seem to do what had been agreed upon 
Perceived size should always be correct or fail, both at init and perf time. Allocated size should only grow at init time. 

Using the new tabinit (which should only be called at init time) and tabcheck (to be called at perf time), they should 
read (for 1D arrays):

```c
static inline int tabinit(CSOUND *csound, ARRAYDAT *p, int size)
{
    size_t ss;
    if (p->dimensions==0) {
        p->dimensions = 1;
        p->sizes = (int32_t*)csound->Malloc(csound, sizeof(int32_t));
    }
    if (p->data == NULL) {
        CS_VARIABLE* var = p->arrayType->createVariable(csound, NULL);
        p->arrayMemberSize = var->memBlockSize;
        ss = p->arrayMemberSize*size;
        p->data = (MYFLT*)csound->Calloc(csound, ss);
        p->allocated = ss;
    } else if( (ss = p->arrayMemberSize*size) > p->allocated) {
        p->data = (MYFLT*) csound->ReAlloc(csound, p->data, ss);
        p->allocated = ss;
    }
    p->sizes[0] = size;
    p->dimensions = 1;
}


static inline int tabcheck(CSOUND *csound, ARRAYDAT *p, int size, OPDS *q)
{
    if (p->data==NULL || p->dimensions == 0) {
      return csound->PerfError(csound, q, Str("Array not initialised"));
    }
    size_t s = p->arrayMemberSize*size;                                 
    if (s > arr->allocated) {                                             
      return csound->PerfError(csound, q, 
        Str("Array too small (allocated %lu < needed %lu), but can't "
            "allocate during performance pass. Allocate a bigger array at init time"), 
        arr->allocated, s);                               
    }
    p->sizes[0] = size;
    return OK;
}

```

**Comment by jpffitch** [comment]
We never agreed.

⁣Sent from TypeApp ​

**Comment by gesellkammer** [comment]
The array semantic's behind your proposed code is:

* first allocation of an array defines the allocated and perceived size
* a new allocation (even from a previous instrument instance) will have a wrong perceived size
* any modification of perceived size at performance time is not allowed, even if this would not imply an allocation

Why??

**Comment by jpffitch** [comment]
> 
> The array semantic's behind your proposed code is:
>
>  *  first allocation of an array defines the allocated and perceived size

yes

>  *  a new allocation (even from a previous instrument instance) will have a
>     wrong perceived size

No use s main iof perceived time;just I dd not remove it in case it was 
useful.  Allocatiis from a previous instance us a bug which I thiught had 
been fixed

>  *  any modification of perceived size at performance time is not allowed,
>     even if this would not imply an allocation
>

yes
> Why??
>

Clean explicable semantics with no (fre?) surprises.  Also oincorporates 
no memory activity at perf time.

**Comment by gesellkammer** [comment]
The problem with that is:

* If there are two i-time allocations of the same array, the second one has no effect and is silently ignored
* Allocations from previous instances are still around (have not been fixed, there was a discussion about reuse of already allocated memory)
* Why not allow modification of perceived size at performance time if perceived size < allocated memory? What is the point of having perceived size at all in that case?

I think that these changes require a wider discussion regarding the semantics. I am sure that users would find the semantics proposed here rather unexplicable.

**Comment by kunstmusik** [comment]
I think the semantics worked out between me and @gesellkammer made sense.  That said, the easiest thing to do at this point is not guess but make unit tests that illustrate the various combinations of operations expected between init-time and perf-time array generating/processing opcodes.  That will make it easy to know exactly what is needed and if anything is not accounted for.

**Comment by gesellkammer** [comment]
> That said, the easiest thing to do at this point is not guess but make unit tests that illustrate the various combinations of operations expected between init-time and perf-time array generating/processing opcodes

That is a great idea. It has been already been settled that no allocations can happen at perf time. We only need to agree on the following:

* It must be decided whether allocations made by a previous instance of an instrument should be reused or discarded 
* It is possible to have multiple init functions acting on the same array. Each init func should grow the array if necessary, and always set its perceived size
* It is possible to modify the perceived size at perf time. There are cases where the resulting size of the array depends on k-values.
* Every array opcode operating at perf time must ensure (not only check) its perceived size (it might have been changed by another opcode)
* Perceived size can't be larger than allocated memory and any attempt to grow an array over its allocated size should result in an error.

**Comment by jpffitch** [comment]
On Mon, 20 May 2019, Eduardo Moguillansky wrote:

>
>       That said, the easiest thing to do at this point is not guess but
>       make unit tests that illustrate the various combinations of
>       operations expected between init-time and perf-time array
>       generating/processing opcodes
> 
> That is a great idea. It has been already been settled that no allocations can
> happen at perf time. We only need to agree on the following:
>
>  *  It must be decided whether allocations made by a previous instance of an
>     instrument should be reused or discarded
>  *  It is possible to have multiple init functions acting on the same array.
>     Each init func should grow the array if necessary, and always set its
>     perceived size

Do not agree

>  *  It is possible to modify the perceived size at perf time. There are cases
>     where the resulting size of the array depends on k-values.

Not agree

>  *  Every array opcode operating at perf time must ensure (not 
only check) its
>     perceived size (it might have been changed by another opcode)

not ncessary

>  *  Perceived size can't be larger than allocated memory and any attempt to
>     grow an array over its allocated size should result in an error.

not appliable

**Comment by vlazzarini** [comment]
Here are some of my thoughts on the subject:

> > * It must be decided whether allocations made by a previous instance of an
> > instrument should be reused or discarded

The best way to do that is to use the AuxAlloc mechanism which is designed
to do this type of thing. I have noted in the past that using Malloc etc may not 
be ideal as they are best designed for more persistent uses. 
This should help avoid undue memory usage.

> > * It is possible to have multiple init functions acting on the same array.
> > Each init func should grow the array if necessary, and always set its
> > perceived size

That is a possibility that we need to examine in terms of use cases and
see whether we permit it or not. I see that if we init an array with size X
and then call fillarray with X+Y inputs we either grow the array or throw an
error. We need to know what is the best of the two.

> > * It is possible to modify the perceived size at perf time. There are cases
> > where the resulting size of the array depends on k-values.

Again, use cases are required for us to make a decision. In this particular
situation, it makes the system more complex, as if we say yes then
the following point is necessary:

> > * Every array opcode operating at perf time must ensure (not only check) its
> > perceived size (it might have been changed by another opcode)

Needed or not depending on the decision relating to previous point.

> > * Perceived size can't be larger than allocated memory and any attempt to
> > grow an array over its allocated size should result in an error.

This again depends on whether size may change at k-time.

So, if you ask my opinion, here’s what I think:

- i-time size changes should be allowed as they do not seem to 
complicate things too much, and would avoid throwing errors
even if at cost of sloppy user code.

- perf-time size changes should preferably be prevented to
avoid making the system overcomplicated. However if there are
use cases that overwhelmingly demonstrate that this is a 
significant limitation, then we need to allow it.

In general, I am easy going with these proposals in any of the
forms presented, as long as we avoid allocating resources at
perf-time.

I also think we should consider modifying the allocation code to use 
AuxAlloc. That should bring the code in line with other types of
variable allocation (e.g. f-sigs). In fact, we should consider that
for strings as well.

========================
Prof. Victor Lazzarini
Dean of Arts, Celtic Studies, and Philosophy,
Maynooth University,
Maynooth, Co Kildare, Ireland
Tel: 00 353 7086936
Fax: 00 353 1 7086952

**Comment by gesellkammer** [comment]
I hope that we agree that previous instances of an instrument should not have any effect in an array's allocated / perceived size (which would be the case if the code here is given green light). 

I circumscribe the use cases here to changes in **perceived** size of an array at k-time

## Array arithmetic

I often use arrays to process frequencies/amplitudes of partials at k-time. Since the amount of partials is not constant in time, the array of frequencies needs to be modified accordingly. If the perceived size is not allowed to be reduced at k-time, then all operations over these arrays affect all elements, making operations much more costly than needed. It is currently possible to use `trim` to set the size at k-time, but according to the new semantics, such an opcode should be banned since it modifies the perceived size during performance.

## Slicearray / trim / genarray

When slicing an array, even if in the current form start and end are i-time variables only, the size of the resulting array is not set accordingly, and currently needs `trim` to set it at k-time. Anyone learning csound would trip over the quirks of using slicearray, trim, genarray, fillarray and lenarray with preallocated or not preallocated arrays, etc. I would much prefer that all these operations set the **perceived** size of the array to the expected value, as long as this value is within the allocated size of the array.

**Comment by jpffitch** [comment]
Eduardo
The comments here are hard to follow.  Any chance of you writing the code you want or givig a detailed plan?  I do not know where to start to implement your scheme.

**Comment by jpffitch** [comment]
i have taken the tabinit and tabcheck functions from above (fixing the error in tabinit) but there are many places where tabcheck is needed but is not used yet.

About to do a lot of testing.

**Comment by jpffitch** [comment]
There is a branch changed-arrays which uses tabinit and tabcheck together with tabinit_like.  Needs tabcheck_like to complete.  Not all checks there yet but until someone agrees this is correct semantics I am loath to spend too much effort on it.

**Comment by gesellkammer** [comment]
Semantics are correct now! Thanks!
